arch/xtensa/core/xtensa_mmu.c

```diff
@@ -244,8 +244,7 @@ static inline uint32_t *alloc_l2_table(void)
 static ALWAYS_INLINE void switch_page_tables(uint32_t *ptables, bool dtlb_inv, bool cache_inv)
 {
 	if (cache_inv) {
-		sys_cache_data_invd_range((void *)ptables, XTENSA_L1_PAGE_TABLE_SIZE);
-		sys_cache_data_invd_range((void *)l2_page_tables, sizeof(l2_page_tables));
+		sys_cache_data_flush_and_invd_all();
```

Comment:
Don't mind the "flush cache on page table change" as an implementation choice.  But I don't understand the commit message about way size aliasing with pages?  You can't have data from alien memory maps in the cache, period.  It's just a correctness thing, has nothing to do with what the address bits mean.

That said: I believe Xtensa has an ASID field in the TLB to allow the OS to elide exactly this flush.  Would be very good to support that instead.  But flush-the-world is fine as an intermediate step.

But can we clarify the commit message, because I'm really kinda confused.

Reply:
There are many things here, let me try to answer one by one:

- Regarding ASID, the way it works is, we set RASID with ASID for all four rings. When the hardware populate the TLB, it gets the RING from the page and then looks the correspondent ASID (from RASID) and use it on TLB. That is one reason to flush the cache  for a  page, otherwise the cached page could have a wrong RING/value/permission. Only flushing the TLB does not solve the problem because the TLB could be re-populated with stale data. That does not explain the aliasing issue though.

- About the aliasing problem, that was trick and I noticed it only later. The HW that I am using has 16kb in each way. So we need 14 bits to address everything, while the page size we are using is 4kb, it means that the page offset is 12 bits. In this situation, a physical memory mapped in two different virtual address may end up in two different cache entries. So, why this is a problem here, if we basically do 1:1 map and since we invalidating the TLB , a miss there would make it read from the physical memory ? The answer is:
  - The page table itself leaves in the virtual memory, since it has to be a contiguous 4MB block, we make a 2 level page table that leaves in the virtual memory and only few pages are really mapped. So, because of this the physical memory for the page table is mapped twice, the correspondent virtual address to its physical address ans this arbitrary virtual address of the page tables (that is the one used by the MMU). This means that a page table entry may have two entries in the cache.
  - Now why flushing the whole TLB does not solve the problem ? The problem here is that the MMU has auto-refill, so, when we have a TLB miss the hw automatically (without any exception) try to load the right page and if this page is on cache it gets from there and remember, a page entry may be mapped in two different virtual address and consequently have two entries in the cache. If one of them is stale we may have a problem

That's why we have cache aliasing and it is a problem. Obviously if the target has a smaller cache or uses a page size bigger than the cache size this flush is not needed ... We can make it build option, but without other uses for now, this additional complexity does not make much sense. The other option is having it more elaborated coloring the cache ... which I am thinking to do later.

One more thing, I tried to manually track possible duplicated entries in the cache but I end up with a lot of checks and conditions and occasional bugs :(

Reply:
I'm still confused: I'm gathering (and maybe I'm wrong?) that you're trying to handle the case where you have a single physical page of entries in your page table **itself** mapped into multiple locations in the table!?  Yikes?  I mean I guess that probably works from a hardware perspective, but just... why?!  Just duplicate the entries, it's only a tiny overhead.

Alternatively: why is the page table living in the dcache anyway?  By definition this is read (on performance paths) only by the hardware itself, and the hardware has an entirely separate cache (the TLB) to optimize repeated reads.  Just map the page tables uncached and this whole problem goes away, no?

Reply:
Imagine you have a page frame that is located in phys mem 0x00000000 and we set the ptevaddr to 0x20000000. This page frame will be mapped in two virtual addresses, 0x00000000 and 0x20000000. Note that the virtual address for the whole page table is mapped (0x20000000 ... 0x20400000).

So, while removing cache for that range and just having the cached for the 1:1 would *probably* avoid the need for flushing the whole cache, there is no restriction (AFAIK) to avoid someone from mapping a physical address in different virtual addresses.

Reply:
> (like, you want the page tables in userspace mappings to have a different address than their physical location --- why?!)

Actually this comment is wrong.  Page tables are inherently virtual in Xtensa so you can have holes & whatnot.  But the point about caching remains the same.  These just don't belong in the dcache, period.

Reply:
This possibly solve the page table, but there is nothing forbidden mapping a physical address in multiple virtual address that is the root cause of the problem and why the flush is being done.